python/cudf/cudf/pandas/proxy_base.py

```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import cupy as cp
+import numpy as np
+
+
+class ProxyNDarrayBase(np.ndarray):
+    def __new__(cls, arr):
+        if isinstance(arr, cp.ndarray):
+            obj = np.asarray(arr.get()).view(cls)
+            return obj
+        elif isinstance(arr, np.ndarray):
+            obj = np.asarray(arr).view(cls)
+            return obj
+        else:
+            raise TypeError(
+                "Unsupported array type. Must be numpy.ndarray or cupy.ndarray"
+            )
+
+    def __array_finalize__(self, obj):
+        self._fsproxy_wrapped = getattr(obj, "_fsproxy_wrapped", None)
+
+    def __array_ufunc__(self, *args, **kwargs):
+        args = (args[0], args[1], np.asarray(args[2]), np.asarray(args[3]))
```

Comment:
Would the `np.asarray` calls effectively make `__array_ufunc__` always take the slow path? Since cupy arrays implement `__array_ufunc__`, I would expect e.g. `np.sum(proxy_array)` to try to execute on cupy first

Reply:
Yes I would for a ufunc like `np.add` or `np.sqrt`. Thanks for mentioning this because I need to do a quick round of testing to make ensure custom ufuncs (written in python) fallback as expected.

Reply:
Okay custom ufuncs seem to be working as expected.

Reply:
Cool thanks for checking